### PR TITLE
Simplify procedure types/overloads

### LINKED ISSUE #949: `inferProcedureInput` broken in published package
This is a super weird one. It may have been me who broke it, and I haven't been able to figure out how/why yet. My team's project hits this error trying to update to v9.6.0 (from v9.0.0-alpha.0, which was before #915 went in but after/somewhere in the middle of #718).

We're finding that *under certain conditions* `inferProcedureInput<>` is returning `unknown` (in fact, it's the type inference of router definition - the procedure itself is `Procedure<Context, Context, unknown, { ...result... }>`). Repro (note - this does _not_ repro in this repo, if you do `import * as trpc from '../src'` in a test, which is what I tried first. You can use this [playground link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgUwB5mQYxgFQJ4YDyAZnAL5zFQQhwDkaG2AtDAcnQNwBQokscAFRwAhgGc4MKGEyVqtOgAEpMgPRjkUAG6a6vcNHgIAXhSo16xiABM93TBAB2Y+CLBgAShACuMTXABeSWlMADpqX00ACgBKUJBNAHNkKO44ehFrEGBHVToAGjTgmSL08J8-KFjSuHjfERhgJyi6PxcAWXrGpwLEGvScsF8ALjhjUIgAIwArLBgohH70yggIUfGXKBzE2MLl5bIYvf24KGQxCAAbHVHxPEdZBbhB33IYwIA+OAADABIEF4wULEVZkACE32O6UONXiwGs1kuyAA7iIzlE7g84E9HGh4IdPnBcah5jEjtwYjxuGwMHBOjAGk1nIFJOwIKQ3J4KppQgB9azIYh1BndZw8RhzfBEYgAHiKKjCOWImgACtRMMhrN4zgBJRxDGAy+mMpxiADadEy2VybRgxtFdAAuh9uB9YqEYBAAKIAR28IkuUuQJBlSBBazgm225DdlO4EuwQZD8pCoSVqvVmu1yD1BqNXSZ5stWRyqlt9qZTpdsdCjggQM9ACFkABVRwAazryMcsR4QA) to see it going wrong:

```ts
import { expectTypeOf } from 'expect-type';
import * as trpc from '@trpc/server'
import {z} from 'zod'

const appRouter = trpc.router().merge(
  'admin/',
  trpc
    .router()
    .mutation('testMutation', {
      input: z.object({
        foo: z.string(),
      }),
      resolve: async ({ input }) => `${input.foo}!`,
    })
    .middleware(async ({ next }) => next()),
);

type Mutations = typeof appRouter._def.mutations;

expectTypeOf<
  trpc.inferProcedureInput<Mutations['admin/testMutation']>
>().toEqualTypeOf<{ foo: string }>();

expectTypeOf<
  trpc.inferProcedureInput<Mutations['admin/testMutation']>
>().not.toBeUnknown();
```

The two `expectTypeOf` tests fail. The second is just to show that the inferred type is `unknown` and that it isn't a typo/bad test setup.

Update: The tests succeed if you remove the `.middleware(async ({ next }) => next())` line (in this case, the middleware is pointless - we are seeing it for a similar middleware that throws under certain circumstances)

Update 2: It can also be repro'd by doing `import * as trpc from '../dist/trpc-server.cjs'` in this repo, which makes things much easier. Could there be some types that are being stripped at build time, maybe?

### PR DESCRIPTION
**Pull request by mmkal**
~Fixes #949~

Follow-up to #955

This simplifies Procedure down to one type:

- ~`CreateProcedureWithInput`, `CreateProcedureWithoutInput`, `CreateProcedureOptions` -> `CreateProcedureOptions` (a single union type)~
- `ProcedureWithInput`, `ProcedureWithoutInput`, `Procedure` -> `Procedure` (a single non-abstract class which _always_ has a `TInput`. The input just happens to be undefined if it was created without an input parser specified)

~Note: there is one downside to this approach which is that for procedures with no inputs, the `input` prop passed to the resolver is typed as `unknown`, where ideally it'd be `undefined` or `never`. IMHO that's a pretty ok tradeoff given how much this simplifies both `procedure.ts` and `router.ts`.~

### CODE REVIEW

**Review by KATT — CHANGES_REQUESTED** [review]
I really like the change but I don't wanna lose the `undefined` type inference - added extra safeguards for this in #963

File: packages/server/src/router.ts [hunk 1]
```diff
@@ -394,14 +396,11 @@ export class Router<
...
-  public subscription<
-    TPath extends string,
-    TInput,
-    TOutput extends Subscription<unknown>,
-  >(path: TPath, procedure: CreateProcedureOptions<TContext, TInput, TOutput>) {
+
+  public subscription(
+    path: string,
+    procedure: CreateProcedureOptions<TContext, any, any>,
+  ) {
```
Comment (mmkal) [inline packages/server/src/router.ts:403]: Note: for the implementation, the jsdoc and types are ignored so using `any` a bit more so using `any` and not duplicating the beta tag etc. here and in the equivalents for `query` and `mutation`.

**Comment by mmkal** [comment]
@KATT updated, I brought back the overloads on `Router`, but removed the extensions of the `Procedure` class. So a smaller simplification, but it passes the new tests now. I also made procedure without input to have a `TInput` of `undefined | null` instead of `undefined`, since `undefined`-only would be a breaking change (one of the e2e tests caught this). Hopefully this change is ok - since the runtime check is `input != null`, it could be either at runtime so this seems appropriate to me. Up to you though if you think it needs another iteration though.

**Review by KATT — APPROVED** [review]
I moved the `undefined/null` logic into `inferProcedureInput` which feels more correct - let me know what you think @mmkal

**Comment by mmkal** [comment]
LGTM!

**Pull request merged**